src/Orleans/Core/GrainFactory.cs

```diff
@@ -167,6 +179,50 @@ internal Task<TGrainObserverInterface> CreateObjectReference<TGrainObserverInter
             return CreateObjectReferenceImpl<TGrainObserverInterface>(obj);
         }
 
+        private static void RegisterGrainTypes(Type grainInterfaceType)
+        {
+            // Do a one-time discovery of all the types used by this grain interface and if there is a serializer
+            // for it, make sure the type is registered with the serialization manager.
+            GrainRegistrationDictionary.GetOrAdd(
+                grainInterfaceType,
+                type =>
+                {
+                    if (type.IsConstructedGenericType)
+                    {
+                        foreach (var typeParam in type.GetGenericArguments().Where(SerializationManager.HasExternalSerializer))
+                        {
+                            SerializationManager.Register(typeParam);
+                        }
+                    }
+
+                    foreach (var paramType in 
+                        type.GetMethods(BindingFlags.Instance | BindingFlags.Public)
+                        .SelectMany(
+                            mi =>
+                            {
+                                var parameterTypes = mi.GetParameters().Select(pi => pi.ParameterType);
+                                if (mi.ReturnType == typeof(void) || mi.ReturnType == typeof(Task))
+                                {
+                                    return parameterTypes;
+                                }
+
+                                if (mi.ReturnType.IsGenericType && mi.ReturnType.GetGenericTypeDefinition() == typeof(Task<>))
+                                {
+                                    return parameterTypes.Concat(mi.ReturnType.GetGenericArguments());
+                                }
+
+                                return parameterTypes.Concat(new[] { mi.ReturnType });
+                            })
+                        .Distinct()
+                        .Where(SerializationManager.HasExternalSerializer))
+                    {
+                        SerializationManager.Register(paramType);
```

Comment:
This approach wouldn't take derived types into account... Not sure what's the best approach... @sergeybykov what do you think we should do?